### PR TITLE
Add internal map hook

### PR DESCRIPTION
**Pull request by kvalakuzhyzp**
We will be using this to know when to setup the function hook.

We may also need to build a hook on first execution of a memory region for the function hook.
This is because we might know where the pointers for the imported functions are, but we would need to keep track of each of them to know when the relocated value is written. One way of doing this is to hook each of those locations and then redirect hooks to those functions when those pointers are written to.

Another way would be to wait until the main module is executed, which we *assume* would be when the relocations have been finished.

### CODE REVIEW

File: src/zelos/memory.py [hunk 1]
```diff
@@ -540,6 +540,9 @@ def map(
             shared=shared,
             reserve=reserve,
         )
+        hooks = self._hook_manager._get_hooks(HookType.MEMORY.INTERNAL_MAP)
+        for hook in hooks:
+            hook(0, address, size, None)
```
Comment (rcourt) [inline src/zelos/memory.py:545]: Consider returning data in internal map hook, rather than `None`

File: src/zelos/memory.py [hunk 2]
```diff
@@ -572,6 +575,9 @@ def map_file(
             prot=prot,
             shared=shared,
         )
+        hooks = self._hook_manager._get_hooks(HookType.MEMORY.INTERNAL_MAP)
+        for hook in hooks:
+            hook(0, address, size, None)
```
Comment (rcourt) [inline src/zelos/memory.py:580]: Consider returning data in internal map hook, rather than `None`

File: tests/test_api.py [hunk 3]
```diff
@@ -106,6 +113,10 @@ def __init__(self, z):
         self.assertEqual(recorded_reads, {})
         self.assertEqual(recorded_writes, {})
 
+        map_callback.assert_not_called()
+        z.memory.map(0xB0BAF000, 0x1000)
+        map_callback.assert_called_with(z, 0, 0xB0BAF000, 0x1000, None)
```
Comment (rcourt) [inline tests/test_api.py:118]: If you change the above & hooks are called with data (not `None`), update the test here.

File: src/zelos/memory.py (hunk 1 above)
Comment (kvalakuzhyzp) [reply src/zelos/memory.py:549]: Caused some of our LTP tests to fail due to timeouts. Left some comments stating that we won't pass the "value" part of the memory callback due to time costs.

File: src/zelos/memory.py (hunk 2 above)
Comment (kvalakuzhyzp) [reply src/zelos/memory.py:588]: Caused some of our LTP tests to fail due to timeouts. Left some comments stating that we won't pass the "value" part of the memory callback due to time costs.

File: tests/test_api.py (hunk 3 above)
Comment (kvalakuzhyzp) [reply tests/test_api.py:118]: Caused some of our LTP tests to fail due to timeouts. Left some comments stating that we won't pass the "value" part of the memory callback due to time costs.

**Pull request merged**